bin/logging.ts

```diff
@@ -18,9 +23,60 @@ export function setLogThreshold(threshold: LogLevel) {
   logThreshold = threshold;
 }
 
-export function log(level: LogLevel, message: string) {
+export function setGlobalProgressListener(listener: IPublishProgressListener) {
+  globalProgressListener = listener;
+}
+
+// Convert log level to event type
+function logLevelToEventType(level: LogLevel): EventType {
+  switch (level) {
+    case 'error':
+      return EventType.FAIL;
+    case 'verbose':
+      return EventType.DEBUG;
+    default:
+      return EventType.DEBUG;
+  }
+}
+
+export function log(level: LogLevel, message: string, percentComplete?: number) {
   if (LOG_LEVELS[level] >= LOG_LEVELS[logThreshold]) {
-    // eslint-disable-next-line no-console
     console.error(`${level.padEnd(7, ' ')}: ${message}`);
+
+    // Write to progress listener if configured
+    if (globalProgressListener) {
+      const progressEvent: IPublishProgress = {
+        message: `${message}`,
+        percentComplete: percentComplete,
+        abort: () => {},
+      };
+      globalProgressListener.onPublishEvent(logLevelToEventType(level), progressEvent);
+    }
+  }
+}
+
+export class ShellOutputHandler {
+  constructor(private readonly progressListener?: IPublishProgressListener) {}
+
+  public handleOutput(chunk: any, isError: boolean = false) {
+    const text = chunk.toString();
+
+    if (isError) {
+      process.stderr.write(text);
+    } else {
+      process.stdout.write(text);
+    }
```

Comment:
So we are still writing directly to stderr and stdout?

Reply:
Yeah we are doing both for now. I could detect if we are being called directly from publish and ignore printing to stdout but I figured it wouldn't hurt

Reply:
> Anything written to process.stdout or process.stderr needs also be pluggable

That's the whole point of the work: That nothing is indiscriminately written to stdout/stderr anymore.